builtInFunctions/baseAccountFreezer.go

```diff
@@ -0,0 +1,114 @@
+package builtInFunctions
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+var guardianKeyPrefix = []byte(core.ElrondProtectedKeyPrefix + GuardiansKeyIdentifier)
+
+// BaseAccountFreezerArgs is a struct placeholder for
+// all necessary args to create a newBaseAccountFreezer
+type BaseAccountFreezerArgs struct {
+	BlockChainHook BlockChainEpochHook
+	Marshaller     marshal.Marshalizer
+	EpochNotifier  vmcommon.EpochNotifier
+	FuncGasCost    uint64
+}
+
+type baseAccountFreezer struct {
+	marshaller     marshal.Marshalizer
+	blockchainHook BlockChainEpochHook
+
+	mutExecution sync.RWMutex
+	funcGasCost  uint64
+}
+
+func newBaseAccountFreezer(args BaseAccountFreezerArgs) (*baseAccountFreezer, error) {
+	if check.IfNil(args.Marshaller) {
+		return nil, ErrNilMarshaller
+	}
+	if check.IfNil(args.BlockChainHook) {
+		return nil, ErrNilBlockChainHook
+	}
+	if check.IfNil(args.EpochNotifier) {
+		return nil, ErrNilEpochNotifier
+	}
+
+	return &baseAccountFreezer{
+		funcGasCost:    args.FuncGasCost,
+		marshaller:     args.Marshaller,
+		blockchainHook: args.BlockChainHook,
+		mutExecution:   sync.RWMutex{},
+	}, nil
+}
+
+func (baf *baseAccountFreezer) checkArgs(
+	senderAccount,
+	receiverAccount vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+	expectedNoOfArgs uint32,
+) error {
+	if check.IfNil(senderAccount) {
+		return fmt.Errorf("%w for sender", ErrNilUserAccount)
+	}
+	if check.IfNil(receiverAccount) {
+		return fmt.Errorf("%w for receiver", ErrNilUserAccount)
+	}
+	if vmInput == nil {
+		return ErrNilVmInput
+	}
+	if !(bytes.Equal(senderAccount.AddressBytes(), receiverAccount.AddressBytes()) &&
+		bytes.Equal(senderAccount.AddressBytes(), vmInput.CallerAddr)) {
+		return ErrOperationNotPermitted
+	}
+	if vmInput.CallValue == nil {
+		return ErrNilValue
+	}
+	if !isZero(vmInput.CallValue) {
+		return ErrBuiltInFunctionCalledWithValue
+	}
+	if len(vmInput.Arguments) != int(expectedNoOfArgs) {
+		return fmt.Errorf("%w, expected %d, got %d ", ErrInvalidNumberOfArguments, expectedNoOfArgs, len(vmInput.Arguments))
+	}
+	if vmInput.GasProvided < baf.funcGasCost {
+		return ErrNotEnoughGas
+	}
+
+	return nil
+}
+
+func isZero(n *big.Int) bool {
+	return len(n.Bits()) == 0
+}
+
+func (baf *baseAccountFreezer) guardians(account vmcommon.UserAccountHandler) (*Guardians, error) {
```

Comment:
this should not be about FREEZE or UNFREEZE. If a person set a guardian, any action that person takes must be made together with signature from the guardian/guardians - any number of guardians. I think a multisignature design is better here. One transaction with multiple signers - like relayed transaction. MultiSign@Action@Sign1@Sign2..@SignN
We check each signatures and launch the action afterwards.

Reply:
As discussed, we will keep it as it is.